Adding support for defined IDs in `TextControl` component

andrewhayward opened the pull request:
## What?
This PR adds support to the `TextControl` component to accept an optional `id` prop. It currently ignores them, and generates its own.

## Why?
As per #24749, it is sometimes useful to provide components with a defined ID (for example, if you need to reference the control with an `aria-*` attribute elsewhere). Ideally all components would work this way (but that is outside the scope of this patch).

## How?
Currently, [`TextControl`](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/text-control) uses [`useInstanceId`](https://github.com/WordPress/gutenberg/tree/trunk/packages/compose/src/hooks/use-instance-id) to generate an ID. As with other components (see [`SelectControl`](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/select-control/index.tsx#L58), for example), this is now only used if an `id` prop is not passed in. This retains backwards compatibility, and for most use-cases will have minimal impact; IDs will continue to be generated and applied.

## Testing Instructions
Unit tests have been added to verify the expected behaviour:
 - Confirms that IDs are generated and applied, is previously
 - Confirms that ID is used if provided
 - Confirms that the `<label>` receives the ID and has a correct `for` attribute

This can also be confirmed manually:
 - Create an instance of a `TextControl`, providing both `id` and `label` props
 - Check that the `id` and `for` attributes match on the generated `<input>` and `<label>` respectively
 - Additionally, clicking on the `<label>` should put focus on the `<input>`

### Testing Instructions for Keyboard
Although this does not impact first-line keyboard access, similar manual testing can be done using a screen reader:
 - Create an instance of a `TextControl`, providing both `id` and `label` props
 - Navigate to the generated `<label>` and action it (<kbd>⌃ Control</kbd> + <kbd>⌥ Option</kbd> + <kbd>Space</kbd> with VoiceOver, for example); focus should move to the `<input>`
 - Navigate to the generated `<input>` and confirm that it has a label matching the `label` prop

Commit:
Adding support for defined IDs in `TextControl`

This patch adds support for custom IDs in the `TextControl` component.

Currently any passed ID prop is ignored, and a auto-generated value is
used instead.

---
Resolves #24749

chad1008 reviewed (approved):
Pending a changelog entry, this LGTM :rocket: :ship:

mirka commented on packages/components/src/text-control/test/text-control.tsx:
```diff
@@ -0,0 +1,46 @@
+			/>
+		);
+
+		const textbox = screen.getByRole( 'textbox' );
+		const label = screen.getByText( labelValue );
+		expect( textbox ).toHaveAttribute( 'id', label.getAttribute( 'for' ) );
```
Nit: Not that this is inherently wrong, but I find it more idiomatic/robust within the testing-library context to simply assert that the `textbox` is accessibly labeled, rather than assert the implementation details.

```suggestion
		expect( screen.getByRole( 'textbox', { name: labelValue } ) ).toBeVisible();
```

☝️ Something like that. What do you think?

andrewhayward replied:
I don't have much experience with `testing-library`, and I didn't know you could do this! That makes much more sense.

mirka reviewed (commented):
Congrats on your first PR, and thank you for adding tests! 💛

mirka commented on packages/components/src/text-control/index.tsx:
```diff
@@ -16,6 +16,13 @@ import BaseControl from '../base-control';
 import type { WordPressComponentProps } from '../ui/context';
 import type { TextControlProps } from './types';
 
+function useUniqueId( idProp?: string ) {
+	const instanceId = useInstanceId( TextControl );
```
FYI we can pass the id prefix and the `idProp` directly into the `useInstanceId()` function to avoid this custom code ✨

https://github.com/WordPress/gutenberg/blob/10d927c2654cadf850fd1930cd03090cbab1cd0a/packages/compose/src/hooks/use-instance-id/index.ts#L40-L48

andrewhayward replied:
I noticed this, and strongly considered using it. But the `useUniqueId` pattern seemed to be in quite a few places already, which made me second guess myself.

I'm quite happy to apply `useInstanceId` "properly" here though, if that's actually the more appropriate direction.

mirka replied:
> I'm quite happy to apply `useInstanceId` "properly" here though, if that's actually the more appropriate direction.

Yes, I think it is. I have a feeling that the existing `useUniqueId` patterns were written before `useInstanceId` was expanded to take three arguments.

Commit:
Updating tests

Changing the test setup for labels to be more `testing-library` idiomatic.

Commit:
Updating CHANGELOG

Adding line item under Enhancements to include `TextControl` changes.

Commit:
Merge branch 'trunk' into 24749/label_not_connecting_to_input_if_props_contains_an_id

Commit:
Fixing use of `useInstanceId`

Dropping the old `useUniqueId` pattern in favour of the more succinct and complete `useInstanceId` usage.

mirka reviewed (approved):
Great work, thank you! 🎉
